prov/usnic/src/usdf_cq.c

```diff
@@ -916,14 +1010,174 @@ usdf_cq_process_attr(struct fi_cq_attr *attr, struct usdf_domain *udp)
 	return 0;
 }
 
+static int usdf_cq_fd_set_nonblock(int fd)
+{
+	int flags;
+
+	flags = fcntl(fd, F_GETFL, 0);
+	if (flags == -1) {
+		USDF_WARN_SYS(CQ, "fcntl getfl failed[%d]\n", errno);
+		return -errno;
+	}
+
+	if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) < 0) {
+		USDF_WARN_SYS(CQ, "fcntl setfl failed[%d]\n", errno);
+		return -errno;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int usdf_cq_create_fd(struct usdf_cq *cq)
+{
+	int ret;
+
+	ret = usd_ib_cmd_create_comp_channel(cq->cq_domain->dom_dev,
+			&cq->object.fd);
+	if (ret) {
+		USDF_WARN_SYS(CQ, "failed to create comp channel\n");
+		return -FI_EINVAL;
+	}
+
+	USDF_DBG_SYS(CQ, "successfully created comp channel with fd %d\n",
+			cq->object.fd);
+
+	/* Going to need this assuming edge-triggered semantics.
+	 */
+	return usdf_cq_fd_set_nonblock(cq->object.fd);
+}
+
+int usdf_cq_trywait(struct fid *fcq)
+{
+	struct usdf_cq *cq;
+	uint64_t ev;
+	int empty;
+	int ret;
+
+	cq = cq_fidtou(fcq);
+
+	switch (cq->cq_attr.wait_obj) {
+	case FI_WAIT_UNSPEC:
+		return FI_SUCCESS;
+	case FI_WAIT_FD:
+	case FI_WAIT_SET:
+		break;
+	default:
+		USDF_WARN_SYS(CQ, "unsupported wait object type\n");
+		return -FI_EINVAL;
+	}
+
+	while (1) {
+		errno = 0;
+		ret = read(cq->object.fd, &ev, sizeof(ev));
+		if (ret == 0)
+			break;
+
+		if (ret < 0) {
+			if (errno == EAGAIN)
+				break;
+			else
+				return -errno;
+		}
+	}
+
+	if (cq->is_soft) {
+		empty = usdf_check_empty_soft_cq(cq);
+	} else {
+		usd_poll_req_notify(cq->c.hard.cq_cq);
+		empty = usdf_check_empty_hard_cq(cq);
+	}
+
+	if (empty)
+		return FI_SUCCESS;
+
+	return -FI_EAGAIN;
+}
+
+
+static int usdf_cq_bind_wait(struct usdf_cq *cq)
+{
+	int ret;
+	struct usdf_wait *wait_priv;
+	struct epoll_event event = {0};
+
+	if (!cq->cq_attr.wait_set) {
+		USDF_DBG_SYS(CQ, "can't unbind from non-existent wait set\n");
+		return -FI_EINVAL;
+	}
+
+	wait_priv = wait_ftou(cq->cq_attr.wait_set);
+
+	event.data.ptr = cq;
+	event.events = EPOLLIN;
+
+	ret = fid_list_insert(&wait_priv->list, &wait_priv->lock,
+			&cq->cq_fid.fid);
+	if (ret) {
+		USDF_WARN_SYS(CQ,
+				"failed to associate cq with wait fid list\n");
+		return ret;
+	}
+
+	ret = epoll_ctl(wait_priv->object.epfd, EPOLL_CTL_ADD, cq->object.fd,
+			&event);
+	if (ret) {
+		USDF_WARN_SYS(CQ, "failed to associate FD with wait set\n");
+		goto err;
+	}
+
+	USDF_DBG_SYS(CQ, "associated CQ FD %d with epoll FD %d using fid %p\n",
+			cq->object.fd, wait_priv->object.epfd, &cq->cq_fid.fid);
```

Comment:
> Should probably put a comment in this routine somewhere saying that the wait set ref count was already incremented in cq_open (there's a comment in there explaining why), just so that future maintainers don't wonder why unbind decrements the refcount, but we don't increment the refcount here in bind.

Sure I can do that.